packages/synapse-react-client/src/components/EntityAclEditor/EntityAclEditor.tsx

```diff
@@ -0,0 +1,352 @@
+    }
+    resetDirtyState()
+  }, [
+    originalIsInherited,
+    originalResourceAccess,
+    parentResourceAccess,
+    resetDirtyState,
+    setResourceAccessList,
+    updatedIsInherited,
+  ])
+
+  const isPublic = updatedResourceAccessList.some(ra =>
+    [
+      AUTHENTICATED_PRINCIPAL_ID,
+      PUBLIC_PRINCIPAL_ID,
+      ANONYMOUS_PRINCIPAL_ID,
+    ].includes(ra.principalId),
+  )
+
+  const canEditResourceAccess = canEdit
+    ? (resourceAccess: ResourceAccess): boolean => {
+        // Users cannot change their own permissions
+        const isSelf = ownProfile.ownerId === String(resourceAccess.principalId)
+        // Users cannot change permission level for the public group, only add/remove it.
+        // To give the public group DOWNLOAD access, ACT must mark it as anonymous access.
+        const isPublicGroup = resourceAccess.principalId === PUBLIC_PRINCIPAL_ID
+        if (isSelf || isPublicGroup) {
+          return false
+        }
+        return true
+      }
+    : false
+
+  function canDeleteResourceAccess(resourceAccess: ResourceAccess): boolean {
+    // Users cannot delete their own permissions
+    const isSelf = ownProfile.ownerId === String(resourceAccess.principalId)
+    if (isSelf) {
+      return false
+    }
+    return canEdit
+  }
+
+  function getDisplayedPermissionLevelOverride(
+    resourceAccess: ResourceAccess,
+  ): string | undefined {
+    if (resourceAccess.principalId === PUBLIC_PRINCIPAL_ID) {
+      return isOpenData
+        ? permissionLevelToLabel['CAN_DOWNLOAD']
+        : permissionLevelToLabel['CAN_VIEW']
+    }
+    return undefined
+  }
+
+  const {
+    sendNotification,
+    isLoading: isLoadingSendMessageToNewACLUsers,
+    isPending: isPendingSendMessageToNewACLUsers,
+  } = useNotifyNewACLUsers({
+    subject: getSubject(entityBundle.entity.name || ''),
+    body: getBody(ownProfile, entityId),
+    initialResourceAccessList: originalResourceAccess,
+    newResourceAccessList: updatedResourceAccessList,
+  })
+
+  const mutationOptions = {
+    onSuccess: () => {
+      if (notifyNewAdditions) {
+        sendNotification()
+      }
+      onUpdateSuccess()
+    },
+    onError: (e: Error) => {
+      setError(e)
+    },
+  }
+
+  const { mutate: createAcl, isPending: isPendingCreateAcl } =
+    useCreateEntityACL(mutationOptions)
+  const { mutate: updateAcl, isPending: isPendingUpdateAcl } =
+    useUpdateEntityACL(mutationOptions)
+  const { mutate: deleteAcl, isPending: isPendingDeleteAcl } =
+    useDeleteEntityACL(mutationOptions)
+
+  const isPending =
+    isPendingCreateAcl ||
+    isPendingUpdateAcl ||
+    isPendingDeleteAcl ||
+    isPendingSendMessageToNewACLUsers
+
+  const hasAclChanged = useMemo(() => {
+    return (
+      originalIsInherited != updatedIsInherited ||
+      !resourceAccessListIsEqual(
+        originalResourceAccess,
+        updatedResourceAccessList,
+      )
```

Comment:
It would be nice if we had a(n OpenAPI generated?) serialized class/type that used more 'correct' data types, like `Set`s, which wouldn't require creating a utility just to check equality.

I investigated this and the generator we would likely use (`typescript-fetch` from `openapi-generator`) does support this in the latest version, but the Synapse OpenAPI spec does not currently indicate which fields are sets. Here's an issue I created: https://sagebionetworks.jira.com/browse/PLFM-8516